src/App.js

```diff
@@ -1,26 +1,86 @@
 import PubSub from 'pubsub-js';
-import { createBrowserHistory } from 'history';
 import memoize from 'lodash.memoize';
 import pick from 'lodash.pick';
 
 import getQueryParameters from './getQueryParameters';
-import { defaultAuthentication } from './frontendAuthWrapper';
-import validateConfig from './validateConfig';
+import { defaultAuthenticatedUser } from './frontendAuthWrapper';
+import * as handlers from './handlers';
 
 export const APP_TOPIC = 'APP';
+export const APP_BEFORE_INIT = `${APP_TOPIC}.BEFORE_INIT`;
+export const APP_CONFIGURED = `${APP_TOPIC}.CONFIGURED`;
+export const APP_VALIDATED = `${APP_TOPIC}.VALIDATED`;
+export const APP_AUTHENTICATED = `${APP_TOPIC}.AUTHENTICATED`;
+export const APP_I18N_CONFIGURED = `${APP_TOPIC}.I18N_CONFIGURED`;
+export const APP_LOGGING_CONFIGURED = `${APP_TOPIC}.LOGGING_CONFIGURED`;
+export const APP_ANALYTICS_CONFIGURED = `${APP_TOPIC}.ANALYTICS_CONFIGURED`;
+export const APP_BEFORE_READY = `${APP_TOPIC}.BEFORE_READY`;
 export const APP_READY = `${APP_TOPIC}.READY`;
 export const APP_ERROR = `${APP_TOPIC}.ERROR`;
 
 /* eslint no-underscore-dangle: "off" */
 export default class App {
   static _config = null;
   static _apiClient = null;
-  static history = createBrowserHistory();
-  static authentication = defaultAuthentication;
+  static history = null;
+  static authenticatedUser = defaultAuthenticatedUser;
   static getQueryParams = memoize(getQueryParameters);
+  static error = null;
+
+  static async initialize({
+    messages,
+    loggingService,
+    overrideHandlers = {},
+    ...custom
+  }) {
+    try {
+      await this.override(handlers.beforeInit, overrideHandlers.beforeInit);
+      PubSub.publish(APP_BEFORE_INIT);
+
+      this.messages = messages;
+      this.loggingService = loggingService;
+      this.custom = custom;
+
+      // Configuration
+      await this.override(handlers.configuration, overrideHandlers.configuration);
+      PubSub.publish(APP_CONFIGURED);
+
+      // Configuration validation
+      await this.override(handlers.validation, overrideHandlers.validation);
+      PubSub.publish(APP_VALIDATED);
+
+      // Authentication
+      await this.override(handlers.authentication, overrideHandlers.authentication);
+      PubSub.publish(APP_AUTHENTICATED);
+
+      // Internationalization
+      await this.override(handlers.internationalization, overrideHandlers.internationalization);
```

Comment:
Why use the full word here and not in the published event?

Reply:
I think i18n is good here too unless you think it would generate confusion somewhere

Reply:
Oh, good catch.  The inconsistency is a mistake.